toolchain/docs/parse.md

```diff
@@ -0,0 +1,758 @@
+        | x | | i32 |             | y | | 1 |
+        +---+ +-----+             +---+ +---+
+          |      |                  |     |
+          +------+------+           +-----+-----+
+                        |                       |
++-----+               +---+ +---+             +---+
+| var |               | : | | = |             | + |
++-----+               +---+ +---+             +---+
+```
+
+Finally, the `;` is used as the "root" of the variable declaration. It's
+explicitly tracked as the `;` for a variable declaration so that it's
+unambiguously bracketed by `var`.
+
+```ascii
+        +---+ +-----+             +---+ +---+
+        | x | | i32 |             | y | | 1 |
+        +---+ +-----+             +---+ +---+
+          |      |                  |     |
+          +------+------+           +-----+-----+
+                        |                       |
++-----+               +---+ +---+             +---+
+| var |               | : | | = |             | + |
++-----+               +---+ +---+             +---+
+   |                    |     |                 |
+   +--------------------+-----+-----------------+-----+
+                                                      |
+                                                    +---+
+                                                    | ; |
+                                                    +---+
+```
+
+This is the completed parse tree.
+
+In storage, this tree will be flat and in postorder. Because the order hasn't
+changed much from the original code, we can do the reordering for postorder with
+a minimal number of nodes being delayed for later output: it will be linear with
+respect to the depth of the parse tree.
+
+**Tokens**:
+
+```ascii
++-----+ +---+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+
+| var | | x | | : | | i32 | | = | | y | | + | | 1 | | ; |
++-----+ +---+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+
+```
+
+**Parse tree**:
+
+```ascii
+        +---+ +-----+             +---+ +---+
+        | x | | i32 |             | y | | 1 |
+        +---+ +-----+             +---+ +---+
+          |      |                  |     |
+          +------+------+           +-----+-----+
+                        |                       |
++-----+               +---+ +---+             +---+
+| var |               | : | | = |             | + |
++-----+               +---+ +---+             +---+
+   |                    |     |                 |
+   +--------------------+-----+-----------------+-----+
+                                                      |
+                                                    +---+
+                                                    | ; |
+                                                    +---+
+```
+
+**Flattened for storage**:
+
+```ascii
++-----+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+ +---+
+| var | | x | | i32 | | : | | = | | y | | 1 | | + | | ; |
++-----+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+ +---+
```

Comment:
I think it might work better to integrate the labels into the ASCII art. As it is, the boundaries between ASCII blocks (which are visible in the rendered form as well as the markdown) make these feel like three separate figures, and push the eye away from seeing the vertical alignment which is one of the key points of the figure.

```suggestion
```ascii
Tokens:
+-----+ +---+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+
| var | | x | | : | | i32 | | = | | y | | + | | 1 | | ; |
+-----+ +---+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+

Parse tree:
        +---+ +-----+             +---+ +---+
        | x | | i32 |             | y | | 1 |
        +---+ +-----+             +---+ +---+
          |      |                  |     |
          +------+------+           +-----+-----+
                        |                       |
+-----+               +---+ +---+             +---+
| var |               | : | | = |             | + |
+-----+               +---+ +---+             +---+
   |                    |     |                 |
   +--------------------+-----+-----------------+-----+
                                                      |
                                                    +---+
                                                    | ; |
                                                    +---+
Flattened for storage:
+-----+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+ +---+
| var | | x | | i32 | | : | | = | | y | | 1 | | + | | ; |
+-----+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+ +---+
```

I think we should add those embedded captions to the other figures as well (see previous suggestions), so that there's clearer visual continuity between the figures, as per my earlier description of them as frames of animation.

Optional: if it works, it might be worthwhile to make these `<pre>` blocks instead of markdown code blocks, in order to put `<b></b>` tags around the captions.

Reply:
I think I'd like to hear more opinions on this. For me, I find it hard to read so much ascii and get a distinction. The switch in font and the boundaries of the sections is intended to reflect that these are distinctly separate storage locations, even if they are conceptually related.

Reply:
That is, the separation you are bothered by is separation which I am trying to make easy to see. So I think we're at cross purposes here.

Reply:
Updated per discussion, including making nodes equal width.